src_c/geometry.c

```diff
@@ -8,6 +8,113 @@
 
 #define PYGAMEAPI_GEOMETRY_NUMSLOTS 21
 
+static PyObject *
+pg_raycast(PyObject *_null, PyObject *const *args, Py_ssize_t nargs)
+{
+    double start_x, start_y;
+    double end_x, end_y;
+    PyObject **farr;
+    Py_ssize_t col_length;
+    Py_ssize_t loop;
+    double angle;
+    double max_dist;
+
+    if (nargs == 3) {
+        if (!PySequence_Check(args[0]) || !PySequence_Check(args[1]) || !PySequence_FAST_CHECK(args[2])) {
+            return RAISE(PyExc_TypeError, "argument of raycast() must be a sequence");
+        }
+        
+        if (!pg_TwoDoublesFromObj(args[0], &start_x, &start_y)) {
+            return RAISE(PyExc_TypeError, "the starting position requires a pair of floats");
+        }
+        if (!pg_TwoDoublesFromObj(args[1], &end_x, &end_y)) {
+            return RAISE(PyExc_TypeError, "the end position requires a pair of floats");
+        }
+        farr = PySequence_Fast_ITEMS(args[2]);
+        col_length = PySequence_Fast_GET_SIZE(args[2]);
+    }
+    else if (nargs == 4) {
+        if (!PySequence_Check(args[0]) || !PySequence_FAST_CHECK(args[3])) {
+            return RAISE(PyExc_TypeError, "argument of raycast() must be a sequence");
+        }
+
+        if (!pg_TwoDoublesFromObj(args[0], &start_x, &start_y)) {
+            return RAISE(PyExc_TypeError, "the starting position requires a pair of floats");
+        }
+        if (!pg_DoubleFromObj(args[1], &angle)) {
+            return RAISE(PyExc_TypeError, "invalid angle");
+        }
+        if (!pg_DoubleFromObj(args[2], &max_dist)) {
+            return RAISE(PyExc_TypeError, "invalid max distance");
+        }
+        end_x = start_x - cos(angle * PI / 180) * max_dist;
+        end_y = start_y - sin(angle * PI / 180) * max_dist;
+
+        if (!PySequence_FAST_CHECK(args[3])) {
+            return RAISE(PyExc_TypeError, "an internal error has occured");
+        }
+        farr = PySequence_Fast_ITEMS(args[3]);
+        col_length = PySequence_Fast_GET_SIZE(args[3]);
+    }
+    else {
+        return RAISE(PyExc_TypeError,
+                    "invalid number of arguments");
+    }
+
```

Comment:
Another thing you could do is shorten the code a bit by loading the colliders sequence "once", meaning outside of the ifs. This is feasable because the colliders list is always the last parameter passed, therefore you could just do:

```C
    if (!PySequence_FAST_CHECK(args[nargs-1])) {
            return RAISE(PyExc_TypeError, "Colliders parameter must be a sequence");
        }
    farr = PySequence_Fast_ITEMS(args[nargs-1]);
    col_length = PySequence_Fast_GET_SIZE(args[nargs-1]);
```
with this and the removal of the same parts inside the two ifs we save 5 lines.

Reply:
This must be resolved again, as @Emc2356 reintroduced it